linker.go

```diff
@@ -1,57 +1,149 @@
 package ebpf
 
 import (
+	"bytes"
+	"encoding/binary"
 	"fmt"
 
 	"github.com/cilium/ebpf/asm"
+	"github.com/cilium/ebpf/internal/btf"
 )
 
-// link resolves bpf-to-bpf calls.
+// The linker is responsible for resolving bpf-to-bpf calls between programs
+// within an ELF. Each BPF program must be a self-contained binary blob,
+// so when an instruction in one ELF program section wants to jump to
+// a function in another, the linker needs to pull in the bytecode
+// (and BTF info) of the target function and concatenate the instruction
+// streams together.
 //
-// Each library may contain multiple functions / labels, and is only linked
-// if prog references one of these functions.
+// Later on in the pipeline, all call sites are fixed up with relative jumps
+// within this newly-created instruction stream to then finally hand off to
+// the kernel with BPF_PROG_LOAD.
 //
-// Libraries also linked.
-func link(prog *ProgramSpec, libs []*ProgramSpec) error {
-	var (
-		linked  = make(map[*ProgramSpec]bool)
-		pending = []asm.Instructions{prog.Instructions}
-		insns   asm.Instructions
-	)
-	for len(pending) > 0 {
-		insns, pending = pending[0], pending[1:]
-		for _, lib := range libs {
-			if linked[lib] {
+// Each function is denoted by an ELF symbol and the compiler takes care of
+// register setup before each jump instruction.
+
+// findNeighbours finds bpf-to-bpf calls in all given progs.
+// It populates each ProgramSpec's neighbour list with pointers
+// to all other programs referenced by its bytecode.
+func findNeighbours(progs map[string]*ProgramSpec) error {
```

Comment:
> Also, what does the string key represent? The function name?

Yes, but as you can see, the key is ignored in the function. It takes a `map[string]*ProgramSpec` to be conveniently callable from `loadProgramSections()` and not quite sure why this would be an issue.

Reply:
> Do you call them that way because they are laid out next to each other? Aren't these more like dependencies?

The term `dependencies` is vague wrt. whether or not it's also meant to include transient dependencies. 'neighbours' is very clear in this regard, not sure what other term we could come up with for this instead.

Reply:
After discussing offline, we're calling this field `references` instead.